File: Day5/part2/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"regexp"
	"strconv"
	"strings"
)

// Checks if the update is valid according to the ordering rules
func isValidUpdate(update []string, orderingRules map[string][]string) bool {
	for index, page := range update {
		previousPages := update[:index]
		shouldBeLaterPages := orderingRules[page]
		for _, p := range previousPages {
			if contains(shouldBeLaterPages, p) {
				return false
			}
		}
	}
	return true
}

// Corrects an invalid update based on ordering rules
func correctUpdate(update []string, orderingRules map[string][]string) []string {
	correctedUpdate := append([]string{}, update...)

	for !isValidUpdate(correctedUpdate, orderingRules) {
		for index, page := range correctedUpdate {
			previousPages := correctedUpdate[:index]
			shouldBeLaterPages := orderingRules[page]
			for _, p := range previousPages {
				if contains(shouldBeLaterPages, p) {
					firstPageToSwapIndex := indexOf(correctedUpdate, p)
					correctedUpdate = append(correctedUpdate[:firstPageToSwapIndex], correctedUpdate[firstPageToSwapIndex+1:]...)
					correctedUpdate = append(correctedUpdate[:index], append([]string{p}, correctedUpdate[index:]...)...)
					break
				}
			}
		}
	}
	return correctedUpdate
}

// Main function to solve the problem
func solve(inputFile string) string {
	// Read the file
	file, err := os.Open(inputFile)
	if err != nil {
		fmt.Println("Error reading file:", err)
		return ""
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	var content string
	for scanner.Scan() {
		content += scanner.Text() + "\n"
	}

	// Extract ordering rules
	orderingRuleRegex := regexp.MustCompile(`(\d+)\|(\d+)`)
	matches := orderingRuleRegex.FindAllStringSubmatch(content, -1)
	orderingRules := make(map[string][]string)
	for _, match := range matches {
		orderingRules[match[1]] = append(orderingRules[match[1]], match[2])
	}

	// Extract pages to check
	pageCheckRegex := regexp.MustCompile(`\d+(,\d+)+`)
	pageMatches := pageCheckRegex.FindAllString(content, -1)
	var pagesToCheck [][]string
	for _, match := range pageMatches {
		pagesToCheck = append(pagesToCheck, strings.Split(match, ","))
	}

	// Process all invalid updates sequentially and calculate the sum
	sum := 0
	for _, update := range pagesToCheck {
		if !isValidUpdate(update, orderingRules) {
			correctedUpdate := correctUpdate(update, orderingRules)
			midIndex := len(correctedUpdate) / 2
			value, _ := strconv.Atoi(correctedUpdate[midIndex])
			sum += value
		}
	}

	return strconv.Itoa(sum)
}

// Utility function to check if a slice contains a value
func contains(slice []string, value string) bool {
	for _, v := range slice {
		if v == value {
			return true
		}
	}
	return false
}

// Utility function to get the index of a value in a slice
func indexOf(slice []string, value string) int {
	for i, v := range slice {
		if v == value {
			return i
		}
	}
	return -1
}

func main() {
	// Call the solve function with the input file
	result := solve("input.txt")
	fmt.Printf("%s\n", result)
}

File: Day5/part2/main.py
import re


# Checks if the update is valid according to the ordering rules
def is_valid_update(update, ordering_rules):
    for index, page in enumerate(update):
        previous_pages = update[:index]
        should_be_later_pages = ordering_rules.get(page, [])
        for p in previous_pages:
            if p in should_be_later_pages:
                return False
    return True


# Corrects an invalid update based on ordering rules
def correct_update(update, ordering_rules):
    corrected_update = update[:]

    while not is_valid_update(corrected_update, ordering_rules):
        for index, page in enumerate(corrected_update):
            previous_pages = corrected_update[:index]
            should_be_later_pages = ordering_rules.get(page, [])
            for p in previous_pages:
                if p in should_be_later_pages:
                    first_page_to_swap_index = corrected_update.index(p)
                    corrected_update.pop(first_page_to_swap_index)
                    corrected_update.insert(index, p)
                    break
    return corrected_update


# Main function to solve the problem
def solve(input_file):
    # Read the file
    try:
        with open(input_file, 'r') as file:
            content = file.read()
    except FileNotFoundError:
        print(f"Error: File '{input_file}' not found.")
        return ""

    # Extract ordering rules
    ordering_rule_regex = re.compile(r"(\d+)\|(\d+)")
    matches = ordering_rule_regex.findall(content)
    ordering_rules = {}
    for match in matches:
        if match[0] not in ordering_rules:
            ordering_rules[match[0]] = []
        ordering_rules[match[0]].append(match[1])

    # Extract pages to check
    page_check_regex = re.compile(r"\d+(?:,\d+)+")
    page_matches = page_check_regex.findall(content)
    pages_to_check = [match.split(",") for match in page_matches]

    # Process all invalid updates sequentially and calculate the sum
    total_sum = 0
    for update in pages_to_check:
        if not is_valid_update(update, ordering_rules):
            corrected_update = correct_update(update, ordering_rules)
            mid_index = len(corrected_update) // 2
            value = int(corrected_update[mid_index])
            total_sum += value

    return str(total_sum)


if __name__ == "__main__":
    # Call the solve function with the input file
    result = solve("input.txt")
    print(result)
